lib/report.js

```diff
@@ -158,7 +160,117 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.failedTests = [];
+    this.testNumeration = 1;
+    this.message = '';
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  parseTestResults(test, offset = 0) {
+    if (offset === 0) this.testNumeration++;
+    test.results.forEach((res, index) => {
+      const testId = index + 1;
+      if (res.type === 'subtest') {
+        const t = res.test;
+        console.log(`${pad(offset + 2)}# Subtest: ${t.caption}`);
+        this.parseTestResults(t, offset + 2);
+        const prefix = t.success ? 'ok' : 'not ok';
+        console.log(`${pad(offset + 2)}${prefix} ${testId} - ${t.caption}`);
+      } else if (res.success) {
+        this.checkSuccess(res, testId, offset + 2);
+      } else {
+        this.listFailure(res, testId, offset + 2);
+      }
+      this.message = '';
+    });
+    console.log(`${pad(offset + 2)}1..${test.results.length}`);
+  }
+
+  checkSuccess(test, id, offset) {
+    const testPrefix = test.success ? 'ok' : 'not ok';
+    let successfulTest = `${pad(offset)}${testPrefix} ${id}`;
+    if (test.caption !== undefined) successfulTest += ` - ${test.caption}`;
+    console.log(successfulTest);
+  }
+
+  listFailure(test, id, offset) {
+    let message = `${pad(offset)}not ok ${id}\n`;
+    message += `${pad(offset + 2)}---\n`;
+    if (test.message) {
+      message += `${pad(offset + 2)}message: '${test.message}'\n`;
+    }
+    message += `${pad(offset + 2)}severity: `;
+    message += test.severity ? `${test.severity}\n` : 'fail\n';
+    message += `${pad(offset + 2)}data:\n`;
+    message += `${pad(offset + 4)}got:\n`;
+    message += `${pad(offset + 6)}- ${test.expected.toString()}\n`;
```

Comment:
As I mentioned in https://github.com/metarhia/metatests/pull/135#discussion_r264727686, there is no need to call `toString` explicitly since it will be executed as it is located inside template literal.
```suggestion
    message += `${pad(offset + 6)}- ${test.expected}\n`;
```